### PR TITLE
Libcloud 353 implement storage volume methods for openstack

### PR DESCRIPTION
**Pull request by bernieke**
Implements storage volume methods for the openstack driver.

### CODE REVIEW

**Comment by Kami** [comment]
Thanks.

I quickly glanced over it and this patch looks good and fits into the current model.

I still need to do a proper review, but I plan to merge it if there aren't any objections after I perform the review.

@alex is currently also looking into improving block storage support in Libcloud so we should also figure out how this fits into his stuff.

**Comment by alex** [comment]
I think this fits in well with the model we're discussing now on the list, so I'll try to give this a review.

**Comment by Kami** [comment]
@alex Sweet, thanks

File: libcloud/compute/drivers/openstack.py [hunk 1]
```diff
@@ -206,6 +206,55 @@ def list_nodes(self):
         return self._to_nodes(
             self.connection.request('/servers/detail').object)
 
+    def create_volume(self, size, name, location=None, snapshot=None):
```
Comment (alex) [inline libcloud/compute/drivers/openstack.py:213]: This doesn't seem to handle snapshots, so it should raise an error if a snapshot is provided.

File: libcloud/compute/drivers/openstack.py [hunk 2]
```diff
@@ -206,6 +206,55 @@ def list_nodes(self):
...
+                                               },
+                                               'availability_zone': location,
+                                           }
+                                       }).success()
+
+    def destroy_volume(self, volume):
+        return self.connection.request('/os-volumes/%s' % volume.id,
+                                       method='DELETE').success()
+
+    def attach_volume(self, node, volume, device=None):
```
Comment (alex) [inline libcloud/compute/drivers/openstack.py]: What happens when device=None, I think this should document teh default mount location.

File: libcloud/compute/drivers/openstack.py [hunk 3]
```diff
@@ -206,6 +206,55 @@ def list_nodes(self):
...
+            data={
+                'volumeAttachment': {
+                    'volumeId': volume.id,
+                    'device': device,
+                }
+            }).success()
+
+    def detach_volume(self, volume):
+        success = True
+        for attachment in volume.extra['attachments']:
```
Comment (alex) [inline libcloud/compute/drivers/openstack.py:252]: Should there be a way to specify to only detach from some nodes?

File: libcloud/compute/drivers/openstack.py [hunk 4]
```diff
@@ -206,6 +206,55 @@ def list_nodes(self):
...
+            if not self.connection.request(
+                '/servers/%s/os-volume_attachments/%s' %
+                (attachment['serverId'], attachment['id']),
+                method='DELETE').success():
+                success = False
+        return success
```
Comment (alex) [inline libcloud/compute/drivers/openstack.py]: And should this return which nodes failed, or raise an exception with them, or something?

**Comment by alex** [comment]
Besides these comments, this looks good to me.

**Comment by bernieke** [comment]
Awesome.

I'm on vacation next week, but as soon as I get back I'll check the behavior of openstack in these cases, and make the necessary changes.

Thanks for the quick turnaround!

**Comment by bernieke** [comment]
I've addressed your remarks in a second commit: https://github.com/bernieke/libcloud/commit/456261fb379f5629ce0f047a757b8307ce181dca

File: libcloud/compute/drivers/openstack.py [hunk 5]
```diff
@@ -59,6 +59,14 @@
 DEFAULT_API_VERSION = '1.1'
 
 
+class CloudStackException(Exception):
```
Comment (Kami) [inline libcloud/compute/drivers/openstack.py]: Here, this should probably be `OpenStackException` and not `CloudStackException` :)

We recently also added a new `ProviderError` base exception class (https://github.com/apache/libcloud/blob/cb9d374732fca01d488e117221baf1fbf6603032/libcloud/common/types.py#L67) for error like this, so please make sure this class inherits from it.

Comment (bernieke) [inline libcloud/compute/drivers/openstack.py]: I've switched to the ProviderError exception class, and fixed the name of the exception (good catch :))

This is the commit with the change: https://github.com/bernieke/libcloud/commit/c72e6bafc3b75716094454fa404236644d11df20

**Comment by Kami** [comment]
Looks good, please attach patch to your JIRA ticket and I'll get it merged.

**Comment by bernieke** [comment]
commits squashed, patch attached to https://issues.apache.org/jira/browse/LIBCLOUD-353

**Pull request closed without merging**